pkg/restic/repository_manager.go

```diff
@@ -0,0 +1,574 @@
+type Backupper interface {
+	// BackupPodVolumes backs up all annotated volumes in a pod.
+	BackupPodVolumes(ctx context.Context, backup *arkv1api.Backup, pod *corev1api.Pod, log logrus.FieldLogger) error
+}
+
+// Restorer can execute restic restores of volumes in a pod.
+type Restorer interface {
+	// RestorePodVolumes restores all annotated volumes in a pod.
+	RestorePodVolumes(ctx context.Context, restore *arkv1api.Restore, pod *corev1api.Pod, log logrus.FieldLogger) error
+}
+
+type BackendType string
+
+const (
+	AWSBackend   BackendType = "aws"
+	AzureBackend BackendType = "azure"
+	GCPBackend   BackendType = "gcp"
+
+	credsSecret = "restic-credentials"
+)
+
+type repositoryManager struct {
+	objectStore   cloudprovider.ObjectStore
+	bucket        string
+	arkClient     clientset.Interface
+	secretsClient corev1client.SecretInterface
+	log           logrus.FieldLogger
+	repoPrefix    string
+	repoLocker    *repoLocker
+}
+
+// NewRepositoryManager constructs a RepositoryManager.
+func NewRepositoryManager(
+	objectStore cloudprovider.ObjectStore,
+	config arkv1api.ObjectStorageProviderConfig,
+	arkClient clientset.Interface,
+	secretsClient corev1client.SecretInterface,
+	log logrus.FieldLogger,
+) RepositoryManager {
+	rm := &repositoryManager{
+		objectStore:   objectStore,
+		bucket:        config.ResticLocation,
+		arkClient:     arkClient,
+		secretsClient: secretsClient,
+		log:           log,
+		repoLocker:    newRepoLocker(),
+	}
+
+	switch BackendType(config.Name) {
+	case AWSBackend:
+		url := "s3.amazonaws.com"
+
+		// non-AWS, S3-compatible object store
+		if config.Config != nil && config.Config["s3Url"] != "" {
+			url = config.Config["s3Url"]
+		}
+
+		rm.repoPrefix = fmt.Sprintf("s3:%s/%s", url, rm.bucket)
+	case AzureBackend:
+		rm.repoPrefix = "azure:" + rm.bucket + ":"
+	case GCPBackend:
+		rm.repoPrefix = "gs:" + rm.bucket + ":"
+	}
+
+	return rm
+}
+
+func (rm *repositoryManager) BackupPodVolumes(ctx context.Context, backup *arkv1api.Backup, pod *corev1api.Pod, log logrus.FieldLogger) error {
+	// get volumes to backup from pod's annotations
+	volumesToBackup := GetVolumesToBackup(pod)
+	if len(volumesToBackup) == 0 {
+		return nil
+	}
+
+	// ensure a repo exists for the pod's namespace
+	if err := rm.ensureRepo(pod.Namespace); err != nil {
+		return err
+	}
+
+	resultChan := make(chan *arkv1api.PodVolumeBackup)
+
+	informer := arkv1informers.NewFilteredPodVolumeBackupInformer(
+		rm.arkClient,
+		backup.Namespace,
+		0,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		func(opts *metav1.ListOptions) {
+			opts.LabelSelector = fmt.Sprintf("%s=%s", arkv1api.BackupNameLabel, backup.Name)
+		},
+	)
+	informer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(_, obj interface{}) {
+				pvb := obj.(*arkv1api.PodVolumeBackup)
+
+				if pvb.Status.Phase == arkv1api.PodVolumeBackupPhaseCompleted || pvb.Status.Phase == arkv1api.PodVolumeBackupPhaseFailed {
```

Comment:
Can this function be run in parallel for 2 pods for the same backup? We'll likely need to match PVBs for this backup's name/UID.